This PR adds a constrained kinematic control example using CoppeliaSim

juanjqo opened the pull request:
![](https://img.shields.io/badge/Tested%20on%20-CoppeliaSim%204.5.1-green) ![](https://img.shields.io/badge/Tested%20on-Windows%2011%2064bits-blue)

Hi @dqrobotics/developers 

This PR adds a constrained kinematic control example using CoppeliaSim. The main goal is to test some new functionalities added in [97](https://github.com/dqrobotics/matlab/pull/97). Specifically, this method uses/tests the following new methods:
- set_synchronous()
- trigger_next_simulation_step()
- set_joint_target_velocities()
- get_joint_velocities()
- wait_for_simulation_step_to_end()

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/23158313/234257783-44fc4539-f745-4a6b-91d9-fe5b40754668.gif)
Best regards, 

Juancho

juanjqo committed:
This commit adds a new example using vrep to tests the new methods for setting and getting the joint velocities.

juanjqo committed:
Update README.md

juanjqo committed:
[constrained_kinematic_control] Updated the example to test wait_for_simulation_step_to_end().

juanjqo committed:
Merge remote-tracking branch 'origin/new_vrep_example' into new_vrep_example

juanjqo committed:
Merge branch 'dqrobotics:master' into new_vrep_example

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+
+try
+
+jointnames={'LBR4p_joint1','LBR4p_joint2',...
```
Lines not corresponding to set up and do not depend on CoppeliaSim should be moved outside the `try-catch` block. I suggest moving 36--69.

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+    xplane = vi.get_object_pose('target_plane');
+    plane_d = Adsharp(xplane,-k_);
+    controller.set_target_primitive(plane_d);
+    controller.set_stability_threshold(0.00001);
```
This can also be moved outside.

juanjqo replied:
Hi, @bvadorno. The line 
```matlab
xplane = vi.get_object_pose('target_plane');|
``` 
depends on CoppeliaSim. I think it is better to keep it inside the `try-catch` block because it is easy to mistake the name of the object, especially from CoppeliaSim 4.3, where the objects can be [identified](https://www.coppeliarobotics.com/helpFiles/en/accessingSceneObjects.htm) via path and an alias. Futhermore, because the controller definition requires both the robot object, and the desired plane, and both depend on CoppeliaSim, I kept it inside `try-catch` block as well. What do you think?

Outside the `try-catch` block:

```matlab
...
jointnames={'LBR4p_joint1','LBR4p_joint2',...
            'LBR4p_joint3','LBR4p_joint4',...
            'LBR4p_joint5','LBR4p_joint6',...
            'LBR4p_joint7'};

%---------------------Robot definition--------------------------%
LBR4p_DH_theta = [0, 0, 0, 0, 0, 0, 0];
LBR4p_DH_d = [0.200, 0, 0.4, 0, 0.39, 0, 0];
LBR4p_DH_a = [0, 0, 0, 0, 0, 0, 0];
LBR4p_DH_alpha = [pi/2, -pi/2, pi/2, -pi/2, pi/2, -pi/2, 0];
LBR4p_DH_type = double(repmat(DQ_JointType.REVOLUTE,1,7));
LBR4p_DH_matrix = [LBR4p_DH_theta;
                    LBR4p_DH_d;
                    LBR4p_DH_a;
                    LBR4p_DH_alpha
                    LBR4p_DH_type];

robot = DQ_SerialManipulatorDH(LBR4p_DH_matrix);
robot.set_effector(1+0.5*DQ.E*DQ.k*0.07);

n = robot.get_dim_configuration_space();
qmin = [-pi -pi -pi/2 -pi -pi/2 -pi -pi/2]';
qmax = [pi pi pi/2 pi pi/2 pi pi/2]';

try
...
```


Inside:
```matlab
...
try           
    xbase = vi.get_object_pose('LBR4p_joint1');
    robot.set_reference_frame(xbase);
    robot.set_base_frame(xbase);
    vi.set_object_pose('base', xbase);
    %----------------------------------------------------------------
    %---------------------Controller definition----------------------%
    qp_solver = DQ_QuadprogSolver();
    controller = DQ_ClassicQPController(robot, qp_solver);
    controller.set_gain(0.2);
    controller.set_damping(0.05);
    controller.set_control_objective(ControlObjective.DistanceToPlane)
    
    xplane = vi.get_object_pose('target_plane');
    plane_d = Adsharp(xplane,-k_);
    controller.set_target_primitive(plane_d);
    controller.set_stability_threshold(0.00001);
    desired_distance_to_target_plane = 0;
    %----------------------------------------------------------------
    %  Plane constraints
    safe_distance = 0.1;
    wall_plane  = Adsharp(vi.get_object_pose('wall_plane'),  k_); 
    floor_plane = Adsharp(vi.get_object_pose('floor_plane'), k_);
    back_plane  = Adsharp(vi.get_object_pose('back_plane'),  k_);
    planes = {wall_plane, floor_plane, back_plane};
...
```

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+    back_plane  = Adsharp(vi.get_object_pose('back_plane'),  k_);
+    planes = {wall_plane, floor_plane, back_plane};
+
+    vi.trigger_next_simulation_step();
```
I suggest adding a quick comment about the need for triggering the next simulation step.

juanjqo replied:
@bvadorno I added it!

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+    planes = {wall_plane, floor_plane, back_plane};
+
+    vi.trigger_next_simulation_step();
+    q =  vi.get_joint_positions(jointnames);
```
This is done on line 88. Is it necessary to get the joint positions here?

juanjqo replied:
@bvadorno, it is not. I removed unnecessary lines.

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+        b = [dists'; (q-qmin); -(q-qmax)];
+        controller.set_inequality_constraint(A,b);
+        %-------------------------------------------------
+        u = controller.compute_setpoint_control_signal(q, 0);
```
Why is the reference `0` instead of `plane_d`?

juanjqo replied:
Hi @bvadorno. The control objective is `DistanceToPlane`. In this context, `0` is the desired distance to the target plane `plane_d`. Aiming to a more intuitive example, I modified the example to have the following

```matlab
...
desired_distance_to_target_plane = 0;
...
u = controller.compute_setpoint_control_signal(q, ...
                                desired_distance_to_target_plane);
```
What do you think?

bvadorno commented on vrep/constrained_kinematic_control/constrained_kinematic_control.m:
```diff
@@ -0,0 +1,125 @@
+        %-------------------------------------------------
+        u = controller.compute_setpoint_control_signal(q, 0);
+        vi.set_joint_target_velocities(jointnames, u);
+        vi.set_object_pose('effector', robot.fkm(q))
```
Why are lines 104 and 105 necessary? If they are, please add a comment.

juanjqo replied:
@bvadorno line 104 `vi.set_joint_target_velocities(jointnames, u); `is used to set the joint target velocities on CoppeliaSim. This command is required when we are working in synchronous mode and the robot joints are set in velocity control mode. Otherwise, the robot won't move when the simulation is triggering. I added a comment in the example.

On the other hand, line 105 `vi.set_object_pose('effector', robot.fkm(q))` is not necessary. It's just to verify if the kinematic model is matching with the robot model used on CoppeliaSim. This is motivated by [issue85](https://github.com/dqrobotics/matlab/issues/85), in which one of the contributing factors to the user implementation not working is using a kinematic model that does not correspond to the robot used in CoppeliaSim. I added a comment in the example. What do you think?

bvadorno reviewed (changes requested):
Hi @juanjqo, there are a few minor modifications you need to do before approving this pull request.

Many thanks,
Bruno

juanjqo committed:
Merge branch 'dqrobotics:master' into new_vrep_example

juanjqo committed:
[constrained_kinematic_control] Added modifications to comply the Bruno's suggestions

- Moved outside the try-catch block lines that are independent on CoppeliaSim.
- Added comments to explain the need for lines related to some CoppeliaSim commands. Ex: trigger_next_simulation_step, set_joint_target_velocities, set_object_pose.
- Added a new variable to describe better why I'm using 0 instead of plane_d in the line controller/compute_setpoint_control_signal.

juanjqo commented:
> Hi @juanjqo, there are a few minor modifications you need to do before approving this pull request.
> 
> Many thanks, Bruno

Hi @bvadorno, thank you for your review. I'm sorry for my late feedback. I implemented most of your suggestions. In some of them, I'll wait for your feedback again. 

Best regards, 

Juancho

juanjqo committed:
[constrained_kinematic_control] Added the wait_for_simulation_step_to_end() method inside the control loop to ensure the synchronous operation.
